<a href="https://colab.research.google.com/github/nikizadehgfdl/platforms/blob/master/gpu_samples/python/simple%20example%201.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import numpy as np

def pow(a,b,c):
    for i in range(a.size):
        c[i]=a[i] ** b[i]
        

vec_size = 100000000
a = b = np.array(np.random.sample(vec_size), dtype=np.float32)
c = np.zeros(vec_size, dtype=np.float32)

In [5]:
import socket
print(socket.gethostname())
%timeit pow(a, b, c)

d78b22291f4a
1 loop, best of 3: 35.5 s per loop


google colab machine d78b22291f4a 1 loop, best of 3: 35.5 s per loop

gfdl gpu box lscgpu50-d                            31.8 s ± 25.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]:
#Needed to do this for colab 
#!find / -iname 'libdevice'
#!find / -iname 'libnvvm.so'
#import os
#os.environ['NUMBAPRO_LIBDEVICE'] = "/usr/local/cuda-9.2/nvvm/libdevice"
#os.environ['NUMBAPRO_NVVM'] = "/usr/local/cuda-9.2/nvvm/lib64/libnvvm.so"

/usr/local/cuda-9.2/nvvm/libdevice
/usr/local/cuda-9.2/nvvm/lib64/libnvvm.so


In [0]:
from numba import vectorize

@vectorize(['float32(float32, float32)'], target='cuda')
def pow2(a, b):
    return a ** b

In [12]:
%timeit pow2(a,b)

1 loop, best of 3: 272 ms per loop


google colab machine d78b22291f4a :  1 loop, best of 3: 272 ms per loop

gfdl gpu box lscgpu50-d :                                 386 ms ± 77.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)